### DISCUSSION TITLE
Is UseCGroupMemoryLimitForHeap enabled?

### DISCUSSION THREAD

**Discussion by alvarlagerlof** [itzg/docker-minecraft-server#1101] [Q&A] [answered]
I'm having trouble correctly setting memory limits, I believe should be done at the Docker level, and not JVM only. Is the JDK you're using in 
the image using UseCGroupMemoryLimitForHeap? Also, why that version in particular. Why not OpenJDK?

I'm seeing a default max and min of 1G. Can I disable this?

**Comment by itzg** [comment, marked as answer]
All of the base image JVMs are using container-aware memory support (the flag is UseContainerSupport BTW):

```
> docker run --rm --entrypoint bash itzg/minecraft-server -c "java -XX:+PrintFlagsFinal -version|grep -i container"
openjdk version "16.0.2" 2021-07-20
OpenJDK Runtime Environment Temurin-16.0.2+7 (build 16.0.2+7)
OpenJDK 64-Bit Server VM Temurin-16.0.2+7 (build 16.0.2+7, mixed mode, sharing)
     bool PreferContainerQuotaForCPUCount          = true                                      {product} {default}
     bool UseContainerSupport                      = true                                      {product} {default}
```

```
> docker run --rm --entrypoint bash itzg/minecraft-server:java8-multiarch -c "java -XX:+PrintFlagsFinal -version|grep -i container"
openjdk version "1.8.0_302"
OpenJDK Runtime Environment (Temurin)(build 1.8.0_302-b08)
OpenJDK 64-Bit Server VM (Temurin)(build 25.302-b08, mixed mode)
     bool PreferContainerQuotaForCPUCount           = true                                {product}
     bool UseContainerSupport                       = true                                {product}
```

however, what you have noticed is the ultimate problem. I need to adjust the scripts to allow `MEMORY` to be unset and leave the JVM to calculate from the container limits.

I'll finally get started on https://github.com/itzg/docker-minecraft-server/issues/742

**Reply by itzg** [reply]
I have now added a note to the memory section describing how this can now be unset:

https://github.com/itzg/docker-minecraft-server#memory-limit

**Reply by alvarlagerlof** [reply]
Thank you! This made my day. I was afraid you wanted it like that and I'd have to fork.

**Reply by alvarlagerlof** [reply]
Actually now I'm having trouble with my containers crashing far below the memory limit. Often at around 50%.

**Reply by itzg** [reply]
If you're not already, make sure you're adjusting `-XX:MaxRAMPercentage` since it looks like it might still default to 25%.

**Reply by itzg** [reply]
...and it could get close to 50% of container memory with heap at 25% and then the non-heap memory from the JVM.

**Reply by alvarlagerlof** [reply]
> If you're not already, make sure you're adjusting `-XX:MaxRAMPercentage`

How do I set that?

**Reply by itzg** [reply]
Via `JVM_OPTS` https://github.com/itzg/docker-minecraft-server#jvm-options

**Reply by itzg** [reply]
I should probably add that to the docs for un-setting `MEMORY`

**Reply by itzg** [reply]
and from the other discussion I realized that should actually be added to `JVM_XX_OPTS`, such as `-e JVM_XX_OPTS="-XX:MaxRAMPercentage=75"`

**Reply by itzg** [reply]
I added an example to https://github.com/itzg/docker-minecraft-server#memory-limit

**Reply by alvarlagerlof** [reply]
Thanks for the quick reponses!

**Reply by alvarlagerlof** [reply]
![image](https://user-images.githubusercontent.com/14835120/139757579-e0cb980a-0157-4917-9a27-d290edf0d0d5.png)

I'm seeing this (free is host, not container) followed by it attempting to use more memory than it has. Am I missing something?

**Reply by alvarlagerlof** [reply]
Considering those default parameters, `podman run --rm -m 128M -e DEBUG_MEMORY=true -e EULA=TRUE itzg/minecraft-server` seems to output the wrong thing for me. (low mem to demonstrate)

**Reply by itzg** [reply]
It's using `free -m`, which yeah apparently isn't container/cgroup aware. Maybe there's an argument I can add there to tell it to do so.
